File: chain/160相交链表.cpp
#include <iostream>
using namespace std;
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode*p=headA;
        ListNode*q=headB;
        int countA=0;
        int countB=0;
        while(p!=nullptr){
            countA++;
            p=p->next;
        }
        while(q!=nullptr){
            countB++;
            q=q->next;
        }
        cout<<countA<<' '<<countB<<endl;
        if(countA<countB){
            ListNode *temp=headA;
            headA=headB;
            headB=temp;
            int t=countA;
            countA=countB;
            countB=t;
        }
        int sub=countA-countB;
        cout<<sub<<endl;
        p=headA;
        q=headB;
        while(sub>0){
            p=p->next;
            sub--;
        }
        while(countB>0){
            if(p==q&&p!=nullptr)return p;
            p=p->next;
            q=q->next;
            countB--;
        }
        return nullptr;
    }
};

File: chain/160相交链表.py
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        count1 = 0
        count2 = 0
        p = headA
        q = headB
        while p != None:
            count1 += 1
            p = p.next
        while q != None:
            count2 += 1
            q = q.next
        if count1 < count2:
            p = headB
            q = headA
            t = count1
            count1 = count2
            count2 = t
        else:
            p = headA
            q = headB
        for i in range(count1-count2):
            p = p.next
        while p is not None and q is not None:
            if p == q:
                return p
            p = p.next
            q = q.next
        return None
        